lambda/custom/src/handlers/FeedingIntentHandler.ts

```diff
@@ -3,17 +3,107 @@ import {
   getRequestType,
   getIntentName,
   getSlotValue,
+  getDialogState,
+  getRequest,
+  HandlerInput
 } from 'ask-sdk-core';
 
-import { babyBuddy } from '../babybuddy';
+import { IntentRequest, Response } from 'ask-sdk-model';
+
+import { babyBuddy, Child, Timer } from '../babybuddy';
 
 import {
   TimerTypes,
   getTimersForIdentifier,
   getSelectedChild,
-  getResolvedSlotValue,
 } from './helpers';
 
+enum FeedingType {
+  FORMULA = 'formula',
+  BREAST_MILK = 'breast milk',
+  FORTIFIED_BREAST_MILK = 'fortified breast milk'
+}
+
+enum FeedingMethod {
+  BOTTLE = 'bottle',
+  LEFT_BREAST = 'left breast',
+  RIGHT_BREAST = 'right breast',
+  BOTH_BREASTS = 'both breasts'
+}
+
+type HandleDialogFunction = (handlerInput: HandlerInput) => Response;
+type StartFeedingFunction = (selectedChild: Child, selectedChildTimer: Timer | undefined) => Promise<string>;
+type StopFeedingFunction = (handlerInput: HandlerInput, selectedChild: Child, selectedChildTimer: Timer) => Promise<string>;
+
+const handleDialog: HandleDialogFunction = (handlerInput: HandlerInput) => {
+  const type = getSlotValue(handlerInput.requestEnvelope, 'Type');
+
+  if (type === FeedingType.FORMULA) {
+    const request = getRequest<IntentRequest>(handlerInput.requestEnvelope);
+    const intent = request.intent;
+    if (intent && intent.slots) {
+      intent.slots.Method.value = FeedingMethod.BOTTLE;
+    }
+
+    return handlerInput.responseBuilder
+      .addDelegateDirective(intent)
+      .withShouldEndSession(false)
+      .getResponse();
+  }
+
+  return handlerInput.responseBuilder
```

Comment:
This is how we handle dynamic slots.  Intents have what is called a "Dialog Delegation Strategy".  By default, the skill uses an auto delegation strategy where it will continue to ask the user for each slot until all required slot values have been filled.  This works great if you always want all the slots every time.

<img width="325" alt="Screen Shot 2020-08-23 at 1 50 16 PM" src="https://user-images.githubusercontent.com/228338/90986314-94cb1f80-e547-11ea-8eb0-53739e13cd4c.png">

In order to dynamically ask for slots based not the previous slot, we have to disable this behavior on the intent and handle it ourselves in code.  If we return this:

```
return handlerInput.responseBuilder
.addDelegateDirective()
     .getResponse();
```

It is the equivalent of the auto delegation and Alexa will just ask for the next slot.  If we want to do something dynamic, we can intercept Alexa, modify the intent so that a slot value is already filled, and Alexa will automatically not ask for that slot.  So if the feeding type is "formula", it could only have come from a bottle so we just fill in the feeding method and return the new intent.  Alexa will then skip asking for the feeding method and just ask for the number of ounces.

Reply:
Haven’t tested yet but this sounds like it makes sense. I wonder, it is possible to have a “skip word” of some (maybe just “skip”)? E.g. —

User: Add a diaper change?
Alexa: Was it wet?
User: Yes.
Alexa: Was it solid?
User: No.
Alexa: What color was it? (Or whatever the question is)
User: Skip.
Alexa: How much was it?
User: Skip.

So that conversion would make a wet diaper change entry with no color or amount setting.

Could be nice to have that skip word for any optional field.

Reply:
I think using "skip" would be pretty easy but my worry is that it wouldn't be intuitive.  I'll do some research and see how Amazon recommends handling this.

Reply:
@cdubz Thinking about this more... A skip word would be easy to implement for the color but probably not as easy for the amount since the amount expects an ```Amazon.NUMBER``` but maybe I could build that in.  My main concern with this is that the "skip" ability would be somewhat hidden from the user so I would want to figure out an intuitive way to handle that.

Also, I was also thinking that it might be best to just remove the Amount slot altogether from the diaper change.  How often is it even used and how would you even determine the amount?  Would it makes to replace it with a relative size like "small" and "large" instead?

Reply:
@rplankenhorn I'm surprised how rigid Alexa Skills are with this. I did some simple searching also couldn't really find anything useful for optional slots. And I think you have a valid point about it not being necessary intuitive.

What if, for optional fields, we added an intermediary question like -- "Do you want to add other details"? E.g.

```
User: Alexa, tell Baby Buddy to add a diaper change.
Alexa: Was it wet?
User: Yes.
Alexa: Was it solid?
User: No.
Alex: Do you want to add other details?
User: No.
```

and

```
User: Alexa, tell Baby Buddy to add a diaper change.
Alexa: Was it wet?
User: Yes.
Alexa: Was it solid?
User: No.
Alex: Do you want to add other details?
User: Yes.
Alexa: What color was it? (Or whatever the question is)
User: Yellow. (Note: "None" should also be a legitimate option here somehow.)
Alexa: How much was it?
User: 2
```

I'm not sure how often amount is used, but there are definitely use cases for parents who need to keep track of baby's output for various reasons. I suppose we could leave it out for the skill, but I'm inclined to support as much as we can of the regular models.

Reply:
@cdubz I like that and it would be pretty easy to implement.  Basically, we would make all the slots required but if the user answers "no" to ```Do you want to add other details?```, we just fill in data for those values to indicate that the user skipped it.

One thing I will say with the amount field in general is that there is no way to have Alexa record a decimal number.  My wife and I encounter this all the time when we are feeding where our baby either doesn't completely finish a bottle or will need a little something extra so the total ounces would be something like 3.5.  In this case, we just use the skill to log 3 ounces and then when our hands are free, we go to the website on our phones and modify the last feeding to be 3.5 ounces.

I think one thing that might help us with these types of problems is to take a step back and figure out what the purpose of the Alexa skill is.  My thinking is that it isn't meant to replace the web interface, only supplement and make it easier to record activities when your hands are full.  We could go as far as to classify which intents would be used when your hands aren't free vs. free.  For instance, when we are starting a feeding, my wife and I 99% of the time use the skill because we are holding our baby when we need to start it.  But I've noticed that we do close out the feeding using the website probably 30% of the time since our hands are usually free (we put our baby back down for nap time at the end of the feeding).

That being said, I still think what you suggested above is the right way to do this.  I'll play around with it and try to get something working.

Reply:
@rplankenhorn Interesting thoughts... I could see the Alexa skill just focusing on timers in generally for start/stop operations and then maybe for feedings and diaper changes we would just stick to the required fields to cover the simple use cases (e.g. Feeding type and method, Diaper Change wet and solid). Sleep and Tummy Time would be handled fine with timers only, and stats like Weight and Temperature don't really need support.

Reply:
<img width="723" alt="Screen Shot 2020-08-28 at 11 19 53 AM" src="https://user-images.githubusercontent.com/228338/91590088-6753ec80-e920-11ea-9cfe-c5503941ec75.png">

@cdubz This has been implemented and works well.  One thing to note and I think it's an Alexa bug, I had to remove the validations for the color in order to get everything to work for some reason.